### PR TITLE
Contract Expiration Refactor

### PR DESCRIPTION
**Pull request by BenjaminHunt**
# bulkRoleManager
- removes expireContracts command

# transactions
- adds expireContracts command
  - adds transaction log
  - sends message to player who's contract expired to explain their free agency options
- updates player/member/user send message behavior to use the `dmHelper` cog.
- updates cut message embed to be red

# misc updates
- formatting, spacing, typing
- updates misc dm references to use `dmHelper`

### CODE REVIEW

File: match/match.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ class Match(commands.Cog):
     MATCHES_KEY = "Matches"
 
     def __init__(self, bot):
-        self.bot = bot
+        self.bot= bot
```
Comment (blister) [inline match/match.py:23]: Why did you delete a space? What did that space ever do to you? :(

File: transactions/transStringTemplates.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+class TransactionsStringsTemplates:
+    contract_expiration_msg = (
+        "We would like to notify you that your contract with the **{franchise}** (team: {team}) franchise has expired."
+        "\n\n**So what does this mean?**\n\nAs a result, you are now a **Free Agent**, and you may sign to any franchise that "
+        "reaches out with an offer. You may re-sign to the same franchise if they offer you a new contract, or you may sign to "
+        "any other offering franchise of your choice. If you do not accept any offers prior to the deadlines in place, you will "
+        "become a Draft Eligible (DE) player, and will be drafted to a team. For more information, you may use the `{p}dates` "
+        "command in the #bot-input channel."
+
+        "\n\nIf you have any questions or concerns, feel free to reach out to your previous GM ({gm}), or open a mod-mail."
```
Comment (blister) [inline transactions/transStringTemplates.py:11]: We call this ModMail everywhere, not mod-mail. 🤷 Silly and pedantic, perhaps.

File: transactions/transactions.py [hunk 3]
```diff
@@ -1,10 +1,16 @@
+
+from pydoc import describe
```
Comment (blister) [inline transactions/transactions.py:2]: Where is describe being used? Not seeing it anywhere in the code other than this import.

**Review by blister — CHANGES_REQUESTED** [review]

File: match/match.py (hunk 1 above)
Comment (BenjaminHunt) [reply match/match.py:23]: IT WAS A MISTAKE, I DIDN'T MEAN TO 😢

File: transactions/transStringTemplates.py (hunk 2 above)
Comment (BenjaminHunt) [reply transactions/transStringTemplates.py:11]: Not at all, consistency is important!

**Review by blister — APPROVED** [review]

**Comment by xentrick** [comment]
Approve

File: teamManager/teamManager.py [hunk 4]
```diff
@@ -1155,6 +1156,7 @@ def get_current_franchise_role(self, user: discord.Member):
                     return role
             except:
                 continue
+        return None
```
Comment (adam-stover) [inline teamManager/teamManager.py:1159]: why explicit `None` return?

File: teamManager/teamManager.py [hunk 5]
```diff
@@ -1122,6 +1122,7 @@ async def _find_team_name(self, ctx, franchise_role, tier_role):
         for team in teams:
             if await self._roles_for_team(ctx, team) == (franchise_role, tier_role):
                 return team
+        return
```
Comment (adam-stover) [inline teamManager/teamManager.py:1125]: why explicit `return` -- linting?

File: transactions/transStringTemplates.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    contract_expiration_msg = (
+        "We would like to notify you that your contract with the **{franchise}** (team: {team}) franchise has expired."
+        "\n\n**So what does this mean?**\n\nAs a result, you are now a **Free Agent**, and you may sign to any franchise that "
+        "reaches out with an offer. You may re-sign to the same franchise if they offer you a new contract, or you may sign to "
+        "any other offering franchise of your choice. If you do not accept any offers prior to the deadlines in place, you will "
+        "become a Draft Eligible (DE) player, and will be drafted to a team. For more information, you may use the `{p}dates` "
+        "command in the #bot-input channel."
+
+        "\n\nIf you have any questions or concerns, feel free to reach out to your previous GM ({gm}), or open a ModMail."
+    )
```
Comment (adam-stover) [inline transactions/transStringTemplates.py:12]: newline EOF

File: transactions/transactions.py [hunk 7]
```diff
@@ -36,6 +49,92 @@ async def genericAnnounce(self, ctx, *, message):
...
+
+                # get team/franchise info before role removal
+                team = (await self.team_manager_cog.teams_for_user(ctx, member))[0]
+                gm_name = self.team_manager_cog._get_gm_name(franchise_role)
+                # franchise_name = self.team_manager_cog.get_franchise_name_from_role(franchise_role)
+                gm: discord.Member = self.team_manager_cog._find_member_by_name(ctx, gm_name)
+
+                # performs role updates
+                await member.remove_roles(*removable_roles)
+                await member.add_roles(*add_roles)
```
Comment (adam-stover) [inline transactions/transactions.py:110]: can these be done concurrently?

```suggestion
                await asyncio.gather(member.remove_roles(*removable_roles), member.add_roles(*add_roles))
```

File: transactions/transactions.py [hunk 8]
```diff
@@ -36,6 +49,92 @@ async def genericAnnounce(self, ctx, *, message):
...
+
+        roles_to_remove = [
+            self.team_manager_cog._find_role_by_name(ctx, "Draft Eligible"),
+            self.team_manager_cog._find_role_by_name(ctx, self.PERM_FA_ROLE),
+            self.team_manager_cog._find_role_by_name(ctx, "Former Player")
+        ]
+
+        trans_channel : discord.TextChannel = await self._trans_channel(ctx)
+
+        for user in userList:
```
Comment (adam-stover) [inline transactions/transactions.py:72]: might add too much complexity for this pass, but in future maybe we can extract out the code in this block and run the processing for each user concurrently?

File: transactions/transactions.py [hunk 9]
```diff
@@ -430,10 +530,46 @@ async def _send_member_message(self, ctx, member, message):
...
+        if member.nick:
+            name = member.nick
+        else:
+            return "", member.name, ""
+        prefix = name[0:name.index(' | ')] if ' | ' in name else ''
+        if prefix:
+            name = name[name.index(' | ')+3:]
+        player_name = ""
+        awards = ""
+        for char in name[::-1]:
```
Comment (adam-stover) [inline transactions/transactions.py:558]: a little hard to follow what's going on in this section, could you add a brief comment showing an example or explaining? I think that'd help a lot.

File: teamManager/teamManager.py (hunk 4 above)
Comment (BenjaminHunt) [reply teamManager/teamManager.py:1159]: I thought `None` was returned by default, but it isn't. This made things a little more confusing when I was debugging my issues. When I was looking at what a value was, it wasn't showing anything, whereas when I return `None` explicitly, I could see that my value wasn't being properly defined.

Why that was happening is a different question. Answer? I don't know lol.

Comment (BenjaminHunt) [reply teamManager/teamManager.py:1159]: Actually I remember what my issue was, I was trying to get franchise/team info after I removed roles (🤦‍♂️). I updated these to explicitly return `None` to help find out what was happening.

File: transactions/transactions.py (hunk 3 above)
Comment (BenjaminHunt) [reply transactions/transactions.py:2]: Let me _describe_ do you what's going on here. This was definitely an accidental import lmao. Tab complete ends up introducing a bunch of imports unintentionally. This one must have slipped through the cracks.

File: teamManager/teamManager.py (hunk 5 above)
Comment (BenjaminHunt) [reply teamManager/teamManager.py:1125]: I'm not sure. That was in previously, and wasn't my LOC. I suppose it might be a good idea to determine some coding standards one way or another. I'm personally partial to an explicit `return None` for reasons I shared in other comments, but I don't think we need to go back and fix all of them right now.

File: transactions/transactions.py (hunk 7 above)
Comment (BenjaminHunt) [reply transactions/transactions.py:110]: Possibly. What does `.gather` do?

File: transactions/transactions.py (hunk 9 above)
Comment (BenjaminHunt) [reply transactions/transactions.py:558]: Sure thing! The purpose of this function is to extract the `franchise_prefix`, `player_name` and `awards`. 

So league awards is defined somewhere as the stars, trophies, (medals from 1s). So `LEAGUE_AWARDS = [⭐, 🏆, 🥇]`. The purpose of this function is to extract the `franchise_prefix`, `player_name` and `awards`. 

It starts by splitting by the pipe character `|`, which yields the prefix.  `<0> | Nehtaro ⭐` is split into `<0>` and `Nehtaro ⭐`. Awards goes at the end of the name, so it starts by traversing the characters in reverse order and getting all rewards the player may have on their name. `name[::-1]` is what reverses the order of the name string. So `Nehtaro ⭐` becomes `⭐ oratheN`. So it tracks characters until it finds one that isn't defined in `LEAGUE_AWARDS`.

At the end of this, it will return `prefix.strip(), player_name.strip(), awards.strip()`. The strip just removes any excess spaces.

That gives you `<0>`, `Nehtaro`, `⭐`

File: transactions/transactions.py (hunk 8 above)
Comment (BenjaminHunt) [reply transactions/transactions.py:72]: Yeah we can definitely revisit this.

File: transactions/transactions.py (hunk 7 above)
Comment (BenjaminHunt) [reply transactions/transactions.py:110]: Gave it a quick test. It doesn't really work out, I ran into some of the same issues where getting access to team/franchise info didn't work properly. So its better for the roles to be removed then added first, awaiting that behavior.
